Nullable Reference Types

cartermp opened the pull request:
This is a draft of [F# nullable reference types](https://github.com/fsharp/fslang-suggestions/issues/577).

It is based primarily on the [C# proposal](https://github.com/dotnet/csharplang/blob/master/proposals/nullable-reference-types.md), with additional considerations for F# based on existing behavior we have.

I expect a lot of things that are written down to change over time (i.e., a good lot of what I've written may not actually work out well, but we'll see 😄 )

vasily-kirichenko commented on RFCs/FS-1060-nullable-reference-types.md:
````diff
@@ -0,0 +1,823 @@
+type C<'T?>() = class end
+```
+
+Issue: looks horrible with F# optional parameters
````
I find question mark the best option because it's the same as in C# and Kotlin and because it's far shorted than `ty | null`. Yes, it looks horrible with F# optional parameters, but they are used quite rarely and this should not get in the way of accepting `?` syntax.

dsyme replied:
I've no great desire to make either `string | null` or `string?` used heavily in F# code.

dsyme commented on RFCs/FS-1060-nullable-reference-types.md:
````diff
@@ -0,0 +1,823 @@
+If an F# class `B` has nullability as a constraint, then an inheriting class `D` also inherits this constraint, as per existing inheritance rules:
+
+```fsharp
+type B<'T | null>() = class end
````
Per comment above, I don't know what this `type B<'T | null>` means, or at least don't expect to have it in the language.  If, per above, allowing nullable instantiations of generic things is the default, then it's implied.

dsyme commented on RFCs/FS-1060-nullable-reference-types.md:
```diff
@@ -0,0 +1,823 @@
+
+If an unconstrained type parameter comes from an older C# assembly, it is assumed to be nullable, and warnings are given when non-nullability is assumed. If it comes from F# and it does not have the `null` constraint, then it is assumed to be non-nullable. If it does, then it is assumed to be nullable.
+
+Additionally, C# will introduce the `class?` constraint. This syntax sugar will compile into the `class` constraint, but the class itself will be decorated with an attribute that tells consumers that the `'T` is a nullable reference type. This attribute will need to be respected by the F# compiler.
```
When you say "the class itself" I'm not sure what you mean.  Do you mean the generic construct (class, method, struct, whatever) gets an attribute? Or the generic type paramater gets at attribute? This should probably go into the .NET metadata section above as we will need an exact spec of what we're interoperating with

Anyway from the F# perspective this feels the same as `type C<'T when 'T : null>`.

dsyme commented on RFCs/FS-1060-nullable-reference-types.md:
```diff
@@ -0,0 +1,823 @@
+
+### Type inference
+
+Nullability is propagated through type inference:
```
I made some comments above about inference, you might want to move them here

dsyme commented on RFCs/FS-1060-nullable-reference-types.md:
````diff
@@ -0,0 +1,823 @@
+let pipeline = f1 >> f2 >> f3 // WARNING - flowing possible null as input to function that assumes a non-null input
+```
+
+Existing nullability rules for F# types hold (i.e., you cannot suddenly assign `null` to an F# record).
````
I am assuming  `R | null` is now a valid type?  Or is that not the intention?

cartermp replied:
I reworded this. What I mean is that this does imply a way to get past "R does not have null as a proper value"

dsyme commented on RFCs/FS-1060-nullable-reference-types.md:
```diff
@@ -0,0 +1,823 @@
+let f2 s = if s <> "" then "hello" else null
+
+// val f3 : string -> string
+let f3 (s: string) =
```
Looking  at line 402 `WARNING - ...` I think you intended to have f3 be of type "string -> (string | null)"

cartermp replied:
This was a mistype of the type of `pipeline`

dsyme commented on RFCs/FS-1060-nullable-reference-types.md:
```diff
@@ -0,0 +1,823 @@
+    | _ -> s
+
+// val pipeline : string -> (string | null)
+let pipeline = f1 >> f2 >> f3 // WARNING - flowing possible null as input to function that assumes a non-null input
```
I expect nullness will be checked through standard type inference.  So the warning message would be more likely to be something akin to "warning: nullness mismatch.  The type ... did not match the type ...`

dsyme commented on RFCs/FS-1060-nullable-reference-types.md:
````diff
@@ -0,0 +1,823 @@
+F# scripts will now also give warnings when using the compiler that implements this feature. This also necessitates a new feature for F# scripting that allows you to opt-out of any nullability warnings. Example:
+
+```fsharp
+#nullability "false"
````
Probably just the existing

    #nowarn "nullness"

where "nullness" is a package of nowarns

dsyme commented on RFCs/FS-1060-nullable-reference-types.md:
```diff
@@ -0,0 +1,823 @@
+## Alternatives
+[alternatives]: #alternatives
+
+The primary alternative is to simply not do this.
```
I think there are lots of other alternatives we should list. e.g. foremost is this:

1. Design a fully sound (apart from runtime casts/reflection, akin to F# units of measure) and fully checked non-nullness system for F# (which gives errors rather than warnings), then interoperate with C# (giving warnings around the edges and for interop)

Given that F# is in good shape w.r.t. nullness this is actually not such a silly thing to consider.

Another is

2. Allow code to be generic w.r.t. nullness (i.e. so you can express "if you give me a null in, I might give you a null back; if you give me a non-null in, I will give you a non-null back")

cartermp replied:
Added

dsyme commented on RFCs/FS-1060-nullable-reference-types.md:
````diff
@@ -0,0 +1,823 @@
+```fsharp
+let ns: string? = "hello"
+
+type C<'T?>() = class end
````
As I've mentioned above, I don't see why we need explicit nullness at type parameter declarations. Just change this example.

dsyme commented on RFCs/FS-1060-nullable-reference-types.md:
```diff
@@ -0,0 +1,823 @@
+
+**Note:** C# supports the equivalent of this with local functions. TypeScript does not support this.
+
+#### Asserting non-nullability
```
Please add a section on the interaction with optional parameters.  e.g. can you declare

    member x.M(?arg: (string | null))

and can you declare an optional arg where the optionality is represented by null (likewise Nullable, an existing issue today)

dsyme commented on RFCs/FS-1060-nullable-reference-types.md:
````diff
@@ -0,0 +1,823 @@
+type C<'T?>() = class end
+```
+
+Issue: looks horrible with F# optional parameters
````
I've not looked beyond this point

cartermp replied:
:D

dsyme commented on RFCs/FS-1060-nullable-reference-types.md:
```diff
@@ -0,0 +1,823 @@
+
+The `[<AllowNullLiteral>]` attribute allows you to declare F# reference types as nullable, but this "infects" any place where the type might be used. In practice, F# types are not normally checked for `null`, which could mean that there is a lot of code out there that does not account for `null` when using a type that is decorated with `[<AllowNullLiteral>]`.
+
+Additionally, nullable reference types is a headlining feature of C# 8.0 that will have a dramatic impact on .NET and its ecosystem. It is vital that we interoperate with this change smoothly and can emit the appropriate information so that C# consumers of F# components can consume things as if they were also C# 8.0 components.
```
(I notice some of this was covered much further below)

dsyme commented:
I left a lot of comments, but basically I feel this is roughly in the right zone.

dsyme commented:
Some offline discussion with @cartermp:

* We discussed `[<DefaultValue>]` attributes on fields. "Yes, you should add a TBD for that in the RFC, also for `CLIMutable`, default struct values and `Unchecked.defaultof<_>` if they aren't there already."

* Re DefaultValue - I think I'd be OK with a warning for this case. The `DefaultValue` thing is in practice it's not where I wanted it to be.  The original intent was roughly that `DefaultValue(check=false)` would be like `Unchecked.defaultof<_>` - an "ugly unchecked" construct.  And plain `DefaultValue` would be checked and nice.  But bugs like https://github.com/fsharp/fslang-suggestions/issues/484 mean that the checking is imperfect and in practice `DefaultValue` is a source of unchecked nulls.

* Separately, I suppose the LINQ methods like `x.FirstOrDefault()` will have return type `T | null`?  I need to consider what that means for the type system, e.g. if `T` is instantiated to `int32`, does that mean there is some rule during substitution that `int32 | null` --> `int32`.  Odd....  That's not exactly what I think of when I see `T | null`.  It's almost like it's `T | default` in that case... 

* I did wonder if maybe F# -> F# violations of nullability should be errors, akin to Units of Measure.  i.e. if you declare `let f (x : string | null) : string = x` then you get an error.     With some zero-runtime-cost construct to explicitly cast your way out of this.

vasily-kirichenko commented on RFCs/FS-1060-nullable-reference-types.md:
````diff
@@ -0,0 +1,823 @@
+
+```fsharp
+let len (str: string | null) =
+    if (str <> null && str.Length > 0) then // OK - `str` must be null
````
I think C# is gonna steel the idea from Kotlin as is https://kotlinlang.org/docs/reference/null-safety.html#checking-for-null-in-conditions :) I wrote a bit of Kotlin some time ago and I must say, it works really well there, but still feels a bit hacky, like a work around some problem (no real pattern matching probably?)

cartermp replied:
The issues I have with not supporting this are:

* Failing to do this will result in warnings on perfectly `null`-safe code that users will expect the compiler _not_ to yell at them about
* For cases like the example shown, it _appears_ simple for a compiler to support (even though it's not simple at all), which to me is a signal that it should be considered
* Expectations may already be in place for this kind of support, even if the languages people use as examples are not in the ML family (though this is something people may come to accept)

I agree that it's not a particularly in the F# style of things, but a success metric of this feature is having very, very, very low false negatives. If we emit them _by design_, that will just be frustrating for people.

cartermp commented on RFCs/FS-1060-nullable-reference-types.md:
```diff
@@ -0,0 +1,823 @@
+
+### Metadata
+
+Nullable reference types are emitted as a reference type with an assembly-level attribute in C# 8.0. That is, the following code:
```
The names for everything are not fully established yet. But we do know that C# will _not_ be using a compiler switch for this.

I think I'll move this section down, or at least have something saying that there will be metadata involved for all concerns, then describe it in more detail after the F#-specific concerns. Your questions here are already addressed by the "Interaction model" section and section for generic representations. So this section would summarize what the metadata is and what it is for.

cartermp commented on RFCs/FS-1060-nullable-reference-types.md:
```diff
@@ -0,0 +1,823 @@
+
+This is by far the most common pattern in F# programming. In the previous code sample, `str` is of type `string | null`, but `s` is now of type `string`. This is because we know that based on the `null` has been accounted for by the `null` pattern.
+
+#### Likely supported: isNull and others using null-handling decoration
```
If there is a way to do this without an attribute, I'd be for it. But I think that if we do not ship with the compiler understanding `isNull`, we'll be seen as having a gaping hole in the feature.

cartermp commented on RFCs/FS-1060-nullable-reference-types.md:
````diff
@@ -0,0 +1,823 @@
+    | _ -> str.Length // OK - 'str' is string
+```
+
+A slight riff that is less common is treating `str` differently when in a wildcard pattern's scope. To reach that code path, `str` must indeed by non-null, so this could potentially be supported.
````
My full opinion is that this should not really be supported; that is, you need to give it a name. But I do think that this may be unexpected by users, even if it is consistent with how things are done today. If we don't do this, we may want to consider a special warning suggesting that you give it an alias if you intend on using `str` as if it were a non-nullable string.

cartermp commented on RFCs/FS-1060-nullable-reference-types.md:
````diff
@@ -0,0 +1,823 @@
+
+```fsharp
+let len (str: string | null) =
+    if str = null then
````
This I don't quite follow. There's nothing fundamentally different about `str = null` and `arr.[0] = null`. If you haven't checked for `arr` prior to this, you can still NRE which is perfectly acceptable (but a warning would be given letting you know about that).

If it were generalized to `expr = null`/`expr <> null` then it's still no different, it's just that `expr` needs to be a nullable reference type (or I suppose nullable value type) for the code to make sense. Having the compiler understand this pattern doesn't seem like it's at odds with what you're saying. Am I missing something?

cartermp commented on RFCs/FS-1060-nullable-reference-types.md:
```diff
@@ -0,0 +1,823 @@
+
+    match str with
+    | null -> -1
+    | _ -> doWork() // But after this line is written it's fine?
```
This I am 100% fine with being not supported, as you kind of have a stretch it a bit to expect the compiler to understand the `null`-safe access. And the verdict is still out on expectations. C# will support this with local methods. TypeScript does not support this by design.

cartermp commented on RFCs/FS-1060-nullable-reference-types.md:
```diff
@@ -0,0 +1,823 @@
+
+#### Asserting non-nullability
+
+To get around scenarios where flow analysis cannot establish a non-null situation, a programmer can use `!` to assert non-nullability:
```
I don't think people will find it acceptable that they have two choices with perfectly `null`-safe code today:

* Rewrite it to be in a form the compiler can handle
* Ignore the warning

I'm not comfortable with either.

Re: "flow analysis" - perhaps a better term can be used here. But we already do have analysis of byref-like types that don't feel particularly non-F#-ish to me today. I'd like to leave the door open for this as an approach (or certainly part of an approach) for analyzing nullability in one's code.

cartermp commented on RFCs/FS-1060-nullable-reference-types.md:
```diff
@@ -0,0 +1,823 @@
+
+### Generics
+
+A type parameter is assumed to have non-nullable constraints when declared "normally":
```
Agreed, now that I've thought about this more.

cartermp commented:
@dsyme Addressed all feedback as best as I can. The most notable changes are removing constraint syntax and changing the non-nullability assertion mechanism from a postfix operater to a function, with the postfix operator listed as an alternative below.

cartermp commented on RFCs/FS-1060-nullable-reference-types.md:
```diff
@@ -0,0 +1,823 @@
+A warning is given when casting from `'S[]` to `('T | null)[]` and from `('S | null)[]` to `'T[]`.
+
+A warning is given when casting from `C<'S>` to `C<'T | null>` and from `C<'S | null>` to `C<'T>`.
+
```
Added

jcouv commented on RFCs/FS-1060-nullable-reference-types.md:
```diff
@@ -0,0 +1,1033 @@
+
+### .NET Metadata
+
+To remain backwards-compatible with existing codebases and interoperate with other .NET languages that support nullability as a concept, F# will emit and respect .NET metadata that concers to the following:
```
Typo: concerns

jcouv commented on RFCs/FS-1060-nullable-reference-types.md:
```diff
@@ -0,0 +1,1033 @@
+To remain backwards-compatible with existing codebases and interoperate with other .NET languages that support nullability as a concept, F# will emit and respect .NET metadata that concers to the following:
+
+* If a type is a nullable reference type
+* If an assembly has nullability as a concept (regardless of what it was compiled with)
```
We should discuss this. Latest C# design does not include such an assembly-level marker. We may need to revive it if it's needed.

jcouv commented on RFCs/FS-1060-nullable-reference-types.md:
```diff
@@ -0,0 +1,1033 @@
+
+In other words, if we cannot be sure that a given reference type is non-nullable, we assume it is nullable.
+
+As a final note, this attribute _can_ be abused. For example, if a method is annotated with `[NonNullTypes(true)]`
```
Incomplete sentence?

jcouv commented on RFCs/FS-1060-nullable-reference-types.md:
```diff
@@ -0,0 +1,1033 @@
+
+To remain "true" to the original code, we could conceivably introduce null obliviousness and thus treat the value coming out of a component as either nullable or non-nullable, depending on how we assign it.
+
+However, this is not actually possible in F#, because the majority of F# code uses type inference to infer a type. Null obliviousness would require explicit type annotations to work as we intend it! So we must assume nullability to remain safe.
```
One option would be to propage null-oblivious types through type inference, until an explicit type is encountered (if ever).
We still need to discuss this in C# LDM, but we may have the same situation with `var`: `var x = obliviousValue; var y = MethodWithInference(x); ...`.

jcouv commented on RFCs/FS-1060-nullable-reference-types.md:
```diff
@@ -0,0 +1,1033 @@
+
+#### Null assignment and passing
+
+A warning is given if `null` is assigned to a non-null value or passed as a parameter where a non-null reference type is expected:
```
Do you mean: ... is assigned to a non-null ~~value~~ __variable__ ...?

jcouv replied:
Also, aside from (1) assignment, and (2) argument-passing, there probably also is third case for returning. Or is that part of the "assignment" category?

cartermp replied:
Re: `value` -> `variable` - we use `value` to refer to many things in F#, distinguishing between a value and a mutable value (with immutable the default).

Re: return value - I've added another example in the code sample that shows that if you attempt to assume the result of a function that returns a nullable reference type to a value that is explicitly non-nullable, it's a warning.

jcouv commented on RFCs/FS-1060-nullable-reference-types.md:
````diff
@@ -0,0 +1,1033 @@
+```fsharp
+// Inferred signature:
+//
+// val makeNullIfEmpty : str:string -> string | null
````
```F#
let neverNull (str: string) =
    match str with
    | "" -> ""
    | _ -> makeNullIfEmpty(str)
```
How do I prevent a `string | null` inference (when user knows better than compiler)?
Is there some `!` operator as in C#?

cartermp replied:
Yep, this would require non-nullability assertions. I'll make a note of that in this section.

To prevent this, you'd need to do something like this:

```fsharp
let neverNull (str: string) =
    match str with
    | "" -> ""
    | _ -> 
        makeNullIfEmpty str
         |> Unchecked.notNull
```

Where `Unchecked.notNull` (or whatever the mechanism is, perhaps `!`) will assert non-nullability.

jcouv commented on RFCs/FS-1060-nullable-reference-types.md:
```diff
@@ -0,0 +1,1033 @@
+
+Example: a `.dll` coming from a NuGet package built with C# 7.1 will be interpreted as if all reference types are nullable, including constraints, generics, etc.
+
+This means that warnings will be emitted when `null` is not properly accounted for. Additionally, when older non-F# components are eventually recompiled with C# 8.0, it will likely be an API-breaking change that will force new F# code to go back and do away with any redundant `null` checks.
```
> force new F# code to go back and do away with any redundant `null` checks.

In C#, redundant checks are ok. Consider the scenario where you're implementing a public API consumed by older compilers. Although you declare your method as not accepting `null`, you may still receive `null`.

cartermp replied:
Thanks, updating this to state that the redundancy is fine, but we may wish to consider emitting something letting the user know that it's redundant.

jcouv commented on RFCs/FS-1060-nullable-reference-types.md:
```diff
@@ -0,0 +1,1033 @@
+
+* All reference types that are _not_ declared in F# are nullable.
+* All F#-declared reference types that have `null` as a proper value are treated as if they are nullable.
+* All F#-declared reference types that do not have `null` as a proepr value are treated as non-nullable reference types.
```
Typo: proper

dsyme commented:
Notes from a discussion with @TIHan on possible implementation approaches, and also some discussion about FSHarp.Core.

### General implementation approach:

Null checking is different to byref checking because AFAICS it also involves inference, and in particular inference that changes ("enhances" I suppose) the types reported across assembly boundaries.
For example, the code

```
let s = [ "a"; null ]
```

presumably infers type `string?` and the code

```
let s = [ "a" ]
```
presumably infers type `string`.  This information needs to be propagated both through the checking of the whole file, the assembly and across assembly boundaries


The way we currently manage this is in type checking, so I don't think we can do it as a separate phase.  Well, we could perhaps do it as a separate phase but we'd have to mutate a lot of the objects we've already created, and propagate the results around.

### TAST TType adjustment?

So I _think_ the way to approach it would be to put nullness info (and variables) into the TAST for types.  This is, however, quite invasive on the codebase.  For example we might start with this:

```
| TType_app of TyconRef * Nullness * TypeInst
...
type Nullness = { mutable Info : NullnessInfo }
type NullnessInfo = | Unknown | Null | NotNull
```

One basic problem is that `TType_app` is used in, like, 202 places or more in the code.  Not the end of the world but a lot.  

Another problem is the pickled info format in TastPickle.fs.  We have to keep that format stable (keep reading old formats, don't produce incompatible new formats for DLLs that aren't compiled with the new switch) but still encode the information.  We'll find some way to do that I suppose.

Filling in the NullnessInfo on import from .NET isn't that hard I suppose: if it's coming from a new-switch .NET DLL then fill in NotNull by default etc.

With all that in place, I suppose we could start to implement rules that actually interpret this information, e.g. in ConstraintSolver.fs and TypeRelations.fs.  Basically if NotNull meets Null then emit a warning, and if Unknown meets either then mutate it to whichever.  This has to be done with Undo in ConstraintSolver.fs, and there will be some other subtleties.

So you want to extend each case of TType with `nullness` possibly? Meaning, not only TType_app, but others as well? That's a spec question that comes down to
* Can we have type `(int * int) | null`, i.e. can reference tuples be nullable?  I suppose so.
* Can we have type `(int -> int) | null`, i.e. can function types be nullable? I suppose so
* Can we have type `T | null`.  I've been pushing back strongly on the C# notion that a type variable `T` is by default non-nullable - that makes little sense to me as a default

Separately, I'll mention that there is another alternative to

```
| TType_app of TyconRef * TypeInst * Nullness 
| TType_tuple of TupInfo * TTypes * Nullness 
```

which is to instead use a single new case

```
| TType_nullness of TType * Nullness
``` 

That may well work out more cleanly but we'd have to establish exactly where these can occur and carefully look for recursive descent cases where nullness is relevant.

#### FSharp.Core and F# metadata format

There is this also a  problem about FSharp.Core - will it "use the new features" or not?  That's really a major issue.  If it doesnt, it won't be consumable by downlevel compilers.  If it doesn't, does that effect things badly? e.g. will strings returned by String.replicate be assumed to be nullable?  That would be so wrong....  Or will we have to manually annotate the whole API of FSharp.Core (I guess we will)

We can treat FSharp.Core as a separate design point.  There are several meanings of "use the new features" and we need to clarify that.  And there are several ground rules, such as FSharp.Core remains binary compatible, and preferably continues to have F# metadata consumable by downlevel compilers (part of what it means to be binary compatible in my eyes).

> Yea, I agree with that. Basically, existing code will always have same metadata and will work with older compilers. If you try to use a new feature, well it's undefined.

I think it will be more subtle than that.  FSharp.Core may well be compiled with the feature enabled and get a null-checked interpretation when consumed by a new compiler.  But it must get the existing interpretation when consumed by an old compiler.   Exactly how we do that is TBD.

> I see because if we annotate a bunch of stuff in FSharp.Core, it can change the metadata.

yes

> And older compilers need to work with that

Yes, whatever it means to "annotate the metadata" can't intrude with the functioning of an old compiler.

> Do we have to store it in metadata? I assume an attribute is enough info to know

At least, that's my position for now.

Do we need to store it in metadata?  Well, it feels like we do.  I mean this is type information and flows across assembly boundaries.  But yes, simply emitting identical metadata for FSharp.Core would be ok, and perhaps with an associated extra metadata bloc with the nullness values.  It's hard to rely on attributes as the code to correlate F# TAST stuff to .NET stuff would be really gnarly.

> I see. So anything that is defined in F#, we store it in metadata rather than relying on attributes? We emit attributes, but only for someone on the outside to consume? Is that right?

Yes. F#-to-F# is always mediated via the F#-specific metadata today

cartermp commented:
@dsyme 

>I think it will be more subtle than that. FSharp.Core may well be compiled with the feature enabled and get a null-checked interpretation when consumed by a new compiler. But it must get the existing interpretation when consumed by an old compiler. Exactly how we do that is TBD.

Perhaps I'm missing something, but isn't this how the annotations will work? That is, if a compiler does not understand what these are, it should just ignore them and see reference types as the same reference types it knows.

We'd need to verify this of course, but I would think that selective annotation accomplishes what we want.

cartermp commented:
Notes from meeting with C# team:

* Names for attributes are not yet set in stone (but changes will be recorded in their spec).
* Null obliviousness is the default for non-F# reference types today. Changing this to be nullable may be too invasive for existing code, even though the warnings are valuable.
* `NonNullTypes(true|false)` is not easily controlled with project options or the project file. May want to reconsider statements about configuration given this.
* "Intent warnings" are being used for C#, which we should consider for things like `[<DefaultValue>]` and `[<CLIMutable>]`
* Changing options to warn on `Some null` in source and emit `FSharpOption<'T>` where `T` is treated as non-null makes the type closer to what is intended for non-F# consumers. Could consider this.
* Consider not complaining against redundant `null` checks. For example, checking for `null` as a library author even when types are declared as non-null makes you resilient to poor use of a library.
* If null obliviousness is done, how does that play into tools? C# is considering some additional notation here, though they aren't quite sure how to do it yet. But they will need to show tooltips differently for a value if it is oblivious and then treated as either nullable or non-nullable later.
* Null obliviousness is highly related to "we don't know if this is nullable or not", where the latter is what we'll need to represent in the compiler to properly handle type inference. If null obliviousness is done for F#, is it an additional flavor of reference type? If so, is that flavor a compiler implementation detail, or is it surfaced into signatures? If it's surfaced, how to represent with F# signature files? Should it even be representable?
* Is an IDE action to turn a warning into an error in an ad-hoc way? Not sure yet.
* C# is considering a "meta code" to represent the whole group of nullability/non-nullability warnings so that it can be warnaserror'd as a group. F# may wish to consider this as well.

cartermp commented:
Linking https://github.com/Microsoft/visualfsharp/issues/3108 for posterity

slang25 commented on RFCs/FS-1060-nullable-reference-types.md:
```diff
@@ -0,0 +1,1090 @@
+There will be several attributes C# 8.0 code can emit:
+
+* `[NotNullWhenTrue]` - Indicates a method handles `null` if the result is `true`, e.g., a `TryGetValue` call.
+* `[NotNullWhenFalse]` - Indicates a method handles `null` fi the result is `false`, e.g., a `string.IsNullOrEmpty` call.
```
typo: ~~fi~~ if

slang25 commented on RFCs/FS-1060-nullable-reference-types.md:
```diff
@@ -0,0 +1,1090 @@
+
+As descibed in the [Interaction Model](FS-1060-nullable-reference-types.md#interaction-model), there are certain behaviors involved in how to work with assemblies that may or may not express nullability.
+
+C# 8.0 code will emit the `[NonUllTypes(true|false)]` attribute:
```
typo? ~~`NonUllTypes`~~ -> `NonNullTypes`

slang25 commented on RFCs/FS-1060-nullable-reference-types.md:
````diff
@@ -0,0 +1,1090 @@
+
+```fsharp
+let len (str: string | null) =
+    if (str <> null && str.Length > 0) then // OK - `str` must be null
````
> OK - `str` must be null

correction: OK - `str` must be non-null

dsyme commented on RFCs/FS-1060-nullable-reference-types.md:
```diff
@@ -0,0 +1,1126 @@
+    public sealed class NullableAttribute : Attribute
+    {
+        public NullableAttribute() { }
+        public NullableAttribute(bool[] b) { }
```
What does the `bool[]` mean?  Is that used for nullability annotations for nested generic types?

cartermp replied:
Not sure, actually. @jcouv ?

dsyme commented on RFCs/FS-1060-nullable-reference-types.md:
```diff
@@ -0,0 +1,1126 @@
+
+#### Null obliviousness
+
+Nullability obliviousness is a concept that C# 8.0 has. A null-oblivious type is one that no assumptions can be made about. Once assigned to a nullable or non-nullable variable, it is treated as if it is nullable or non-nullable, respectively.
```
I need more information on this - e.g. examples where would C# infer null-obliviousness.

The natural corresponding thing in F# would be to have nullability inference variables.  For example, the most natural implementation in F# for nullability inference would mean that even the literal "a" would initially be assigned type `string | ?` where the `?` represents an inference variable which can be filled in with values "0" (indicating `string`) or "1" (indicating `string | null`).  Thus "obliviousness" is just a dangling inference variable.

This approach is used partly because we never merge to a "common super type" during type inference.  So when we have

    let f (s: string | null)  = if b then "a" else s

the `then` and `else` branches need the same type.  The type of `s` is `string | null` and the type of the literal `"a"` needs to be compatible with this.

Normally this approach to annotation inference is sufficient for F# (e.g. for units of measure and type inference) and means we won't need extra concepts like "null obliviousness" - they should just drop out naturally.  A good corresponding example is that `1.0<_>` is "unit oblivious")

dsyme replied:
> Reference types we cannot determine to be non-nullable will be assumed to be nullable.**

I'm not sure where "assumed to be nullable" is coming dfrom or why we would do that for F#.  The only things we would assume to be nullable are things coming from .NET assemblies, and even that should be explicit assuming the assembly is annotated.

Note this would still apply even if we use nullability inference.  Nullability inference variables that are not "filled in" by the end of inference would be given a default value.  I believe in all situations for regular F# code that default value would be "non-nullable".

If there are counter examples we should add them

cartermp replied:
The main gist of this (which I'll clarify):

* If C# does not see `NonNullTypes(false)` for a scope, it treats that scope as null-oblivious
* If something cannot be inferred to be nullable or non-nullable (`NonNullTypes(false)`), then **nullabulity or non-nullabulity is not inferred**
* Once a value is assigned to a variable, it is treated as either nullable or non-nullable (e.g., assigning to `string?` or `string`), based on how it is assigned
* The variable is treated as nullable or non-nullable from that point forward
* `var x = ...` is still undecided, since type inference means that you need to make a decision one way or another

cartermp replied:
From the F# perspective, that `var x = ...` case is much more of a problem, since we use type inference everywhere. So I'm leaning towards us not having nullability obliviousness as a concept that we surface in user code.

dsyme commented on RFCs/FS-1060-nullable-reference-types.md:
```diff
@@ -0,0 +1,1126 @@
+
+To remain "true" to the original code, we could conceivably introduce null obliviousness and thus treat the value coming out of a component as either nullable or non-nullable, depending on how we assign it.
+
+However, this is not actually possible in F#, because the majority of F# code uses type inference to infer a type. Null obliviousness would require explicit type annotations to work as we intend it! Failing the propagation of a null-oblivous type through type inference (see [Type Inferece](FS-1060-nullabl-reference-types.md#type-inference)), we must assume nullability to remain safe. (**NOTE:** t ...
```
> Failing the propagation of a null-oblivous type through type inference (see [Type Inferece](FS-1060-nullabl-reference-types.md#type-inference)), we must assume nullability to remain safe. (**NOTE:** this is also under discussion for C# with `var x = ...`).

Per the above, I don't really understand this comment. Perhaps it will become clear later

TIHan commented:
Notes

* Does `x :? (string | null)` generate a warning because the runtime type information doesn't have nullness
* information is dropped in F# quotations


Test cases
* can pass obj to `obj | null`
* can pass string to `obj | null`
* can't pass `obj | null` to `obj`

dsyme commented:
Further notes:

Small interaction with type abbreviations, e.g. can you do

    type S = string | null
    type S2 = S | null

or

    type S = string | null
    let f (x: (S | null)) = ...

dsyme commented:
ANother note: the syntax `string | null` may be tricky to implement.  For example, 

    let f2 (_ : string | null) = 1

is valid code today because it is interpreted as a union pattern match combined with a type-annotated wildcard pattern:

 let f2 ((_ : string) | null) = 1

Alternatives that look simpler to implement are

    let f (x: string?) = x
    let f (x: string or null) = x

and there is also the option of using

    let f (x: Nullable<string>) = x

davidglassborow commented:
@voronoipotato this is covered in the RFC document

slang25 commented:
Specifically [here](https://github.com/fsharp/fslang-design/blob/de0d0fe9099167e31b8eaded1bf37b334ec1df49/RFCs/FS-1060-nullable-reference-types.md#option-or-valueoption)

voronoipotato commented:
Thank you so much guys.

auduchinok commented:
I'd like to bring nullability for tuples into consideration.
When tuples are coming from external non-F# code they can have `null` as a proper value while F# treats them as non-null values making working with them either more error-prone or not very convenient.

An example method from R# SDK I had to deal with recently:
```csharp
[CanBeNull]
Tuple<IProject, TargetFrameworkId> GetProjectAndTargetFrameworkIdByOutputAssembly(FileSystemPath assemblyPath);
```

cartermp commented:
@auduchinok For interop, an analogous situation would be like this: https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQDk4BstWFiAJQkiQDsBjCAFQE8AHFAGhhAEssAfAAQAYACbgEYA3AFgAUH0FCAdIThkYbALYRZAYQD2K+hwgIAygYBubKsnETJAbXRaymHHUbIAFDARwIASgC6ktwAzIIATPwa/ADekvxxgiHUcPQEADzCvEwyvAB8APz8ALJuPtGx8RXcAOz8ZNhYVhUAvpJNQA

Which is sort of like this:

```csharp
[NonNullTypes(true)]
public class C
{
    [return: Nullable(new bool[] {
        true,
        false,
        false
    })]
    public Tuple<string, string> M()
    {
        return null;
    }
}
```

And we'd respect the nullability of the construct. I doubt we'd emit that because we don't have `null` as a proper value, but consumption of such a thing is considered in scope.

dsyme commented:
I adjsuted the syntax to `string?` for now.  `string | null` is very hard to support.

I will pull this and we can continue the discussion

mindlid commented:
@dsyme does the literal types proposal [1195](https://github.com/fsharp/fslang-suggestions/issues/1195), may allow for `string | null` where null would be a literal value type? this is the way it works on typescript unions
